src/System.Private.CoreLib/shared/System/Range.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public readonly struct Range
+    {
+        public Index Start { get; }
+        public Index End { get; }
+
+        private Range(Index start, Index end)
+        {
+            Start = start;
+            End = end;
+        }
+
+        public override bool Equals(object value)
+        {
+            if (value is Range r)
```

Comment:
Last I checked, for value types this pattern actually resulted in worse code than doing `if (value is Range) { Range r = (Range)value; ... }`, but maybe something's changed?

Reply:
Do you mean perf wise it is worse when using (value is Range r)?  I don't know that. 

I can revert it back if @jkotas is ok to do so.

Reply:
I have looked around ... we do not use this anywhere else. Sounds reasonable to revert it

Reply:
> Do you mean perf wise it is worse when using (value is Range r)? I don't know that.

It's possible the JIT fixes it all up to be equivalent, I've not looked, but the IL for `if (value is Range r) { ... }` is a bit worse than for `if (value is Range) { Range r = (Range)value; ... }`.